internal/tfdiags/checks.go

```diff
@@ -0,0 +1,73 @@
+package tfdiags
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/hcl/v2"
+)
+
+var _ Diagnostic = CheckBlockDiagnostic{}
+
+// CheckBlockDiagnostic is a diagnostic produced by a Terraform config Check block.
+//
+// It only ever returns warnings, and will not be consolidated as part of the
+// Diagnostics.ConsolidateWarnings function.
+type CheckBlockDiagnostic struct {
```

Comment:
This new type might be fine if it never ends up needing to implement any of the other diagnostic interfaces in a special way, but we've had challenges with this sort of specialized implementation in the past and so that's what the `ExtraInfo` method was intended to solve, so we can hook additional machine-readable context onto some of the other diagnostic implementations.

For example, HCL's version of `ExtraInfo` (a field in `hcl.Diagnostic`) can carry a `hclsyntax.FunctionCallDiagExtra` value if the error occurred while calling a function, and the UI then uses that to annotate the diagnostic with information about the call.

If this won't ever wrap other diagnostic types that we might need to give special treatment too elsewhere then I'm not too concerned about it, but just wanted to note the other option in case it's more attractive. I think at the moment only `hcl.Diagnostic` can _actually_ carry an "extra info" -- other diagnostic types just always return `nil` -- but we could potentially add it to some of the other types if needed, if check blocks can potentially generate diagnostics that are not `hcl.Diagnostic` values.

Reply:
Here's the existing example of how we use the "function call" annotation that HCL generates:

https://github.com/hashicorp/terraform/blob/c960b16e87dca456f89b27c52f1995e4edab8170/internal/command/views/json/diagnostic.go#L340-L353

Reply:
Some other examples of existing uses of this pattern, in case that helps decide whether the pattern would be appropriate for your new use-case:

https://github.com/hashicorp/terraform/blob/c960b16e87dca456f89b27c52f1995e4edab8170/internal/tfdiags/diagnostic_extra.go#L105-L137

https://github.com/hashicorp/terraform/blob/c960b16e87dca456f89b27c52f1995e4edab8170/internal/tfdiags/diagnostic_extra.go#L139-L171

We use these two to improve the relevance of the extra diagnostic information in the diagnostics UI, by only mentioning that values are unknown or sensitive if we're reporting an error that could potentially have been caused by those conditions. (We found that otherwise users would often assume that the unknown or sensitive value was the caused of the error even when it was irrelevant.)

Reply:
Thanks for this! I did spend some time looking into this but I don't think it will work easily in this case. 

Since we want to wrap all diagnostics from the check blocks, we have a bunch of function calls that return groups of diagnostics and it's too much for us to go through and add the ExtraInfo statement to all of them (as well as passing down that we want the severity to always be Warning).

The other option would be to try and turn the diagnostics we have into concrete types, and then set the extra infos and warnings, or try and work something else.

I could spend more time trying to figure something out, but since for now we don't need to wrap other diagnostic types I feel like we can be lazy and revisit this solution if we ever need to do this instead of spending time now when it might not be necessary.

Reply:
I think what you've done here already will be fine as long as we're careful with how things get nested... the family of diagnostic types that start off without source location information and then get annotated with it later are the main worry I have here, but that can be solved by making sure we always call the function that adds the context before we call the function that wraps with this new type, so that the wrapper doesn't hide the extra interface implemented by the contextual diagnostic types.

Reply:
If we ever needed to, since this is only ever intended to be an outer wrapper, we could add some sort of "unwrap" interface and make `tfdiags` aware of it to strip off and reapply this marker for other operations, but I don't think we need to go that far yet. This should always at the end of the call chain right now and shouldn't interfere with the other diagnostic types.

Reply:
Indeed... this ExtraInfo mechanism already has that kind of unwrapping though, so if that did become necessary I'd suggest that we switch over to that. We don't need to worry about it for now, though.